Hiero import/export plugins

apetrynet opened the pull request:
Initial set of plugins for importing and exporting OTIO files in Hiero (and Nuke Studio). There's a README located under `opentimelineio_contrib/application_plugins/hiero` with instructions on how to install and use.

Plugins have been developed in Hiero12, but should work in earlier versions as well.  Please report if it doesn't and I'll add some version toggles etc. (I have access to Hiero11.x and 12.x.)

There are most likely some bugs and/or questionable decisions here, so please give feedback and report issues as you go and I'll iron them out as quickly as I can.

Features:
-------------
- [x] Multiple tracks
- [x] Tags
- [x] Markers
- [x] Fade in/out and cross dissolves
- [x] Nested sequences are created separately and replaced with gaps in the main sequence.
- [x] Simple re-time (reverese, constant slowdown, constant speedup)
- [x] Add tracks to existing Sequence


Todo:
-------
- [x] Implement `ImageSequenceReference` support
- [x] Pass Sequence rate and Time Code to Timeline on export 
- [x] Assign rate and Time Code from Timeline's global_start_time on Import 
- [x] Confirm Simple re-time
- [x] Look into Nuke Studio integration
- [x] Add OTIO Import to ~~main menu~~ among other places
- [x] Add track(s) from OTIO file in Timeline view (like the Add track form EDL/XML option)

Future:
----------
- [ ] Tests. Hiero doesn't have a command line mode so not sure automated tests are possible.
- [ ] Add Drag&Drop support

Limitations:
---------------
* Tags/markers are applied to both clips and track items on import if no metadata indicates source type.

ssteinbach commented:
@apetrynet is this still WIP?  Just wanted to check in on this and see how it was going.

apetrynet commented:
Hey @ssteinbach !
I haven't had the chance to do any additional coding on this since the PR, but want to produce a Hiero project and/or an XML for a proper round trip test. Not sure if it's possible to make tests for this that run in the coverage tests.
Do you know if anyone else has had a chance to try it out?

apetrynet commented:
@ssteinbach 
I suspect some of the test fail because I haven't added the `application_plugin` folder etc. to `setup.py`, Should I or have you come up with another approach for such plugins?
I still have some work to do with this plugin, but finally got some time to work on it again.

jminor commented:
If you add the `application_plugin` folder to the `check-manifest --ignore` in `tox.ini` that should tell the build that you don't want it in the built package.

apetrynet commented:
OK, thanks!

apetrynet commented:
@jminor Uhm.. I can't seem to find any `tox.ini` or reference of `check-manifest` anywhere. I grep'ed through the whole repo. Is this something set online in travis?

jminor commented:
Ah, yes, this just comes from the fact that the master branch has advanced since you made your feature branch. I suggest rebasing your branch onto the current master - that will make it as though you had branched today, with your changes appended after that. This article is a bit long, but explains it pretty well: https://blog.algolia.com/master-git-rebase/

apetrynet commented:
@jminor Thanks for the link. Turned out I had forgotten to actually merge from my local master to my feature branch so I avoided rebasing..
The state of the plugin is where I really would like someone outside my environment to try it. I have refactored quite a bit and added the following features:
* Markers/Tags (import and export)
* Support for nested sequences. (Hiero doesn't support it so I just create separate sequences)
* Transitions are more robust

ssteinbach commented:
@apetrynet if you're still working on this, thats all good - but it would be cool to get this rebased at least.  Maybe you could add a checklist of stuff you'd want to work on for it before you consider it non-wip?

apetrynet commented:
@ssteinbach Do you know how I can disable CI for py3.x on this one? 
Hiero is still python2

ssteinbach commented:
@apetrynet I think you're looking for `unittest.skipIf`.  See the RV adapter for example:

```python
@unittest.skipIf(
    (sys.version_info > (3, 0)),
    "RV Adapter does not work in python 3."
)
```


https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/234b3f319e4e128674a2074364ae7a1077fbf3db/contrib/opentimelineio_contrib/adapters/tests/test_rvsession.py#L464

apetrynet commented:
Thanks @ssteinbach ! In this case I don't have a unittest and it turned out to be a flake issue I managed to resolve. 
I'll just try to use "future proof" syntax for now.

apetrynet commented:
@hasielhassan I noticed you gave this PR a heart earlier, would you be interested in trying out these plugins? It would be great if someone other than me check if they work :)

apetrynet commented:
@Eiken, I noticed via your PR that you use Hiero. Would you be able to give this plugin a spin?

Eiken commented:
> @Eiken, I noticed via your PR that you use Hiero. Would you be able to give this plugin a spin?

Sure thing, I'll get to it sometime early next week. Bank holiday in Sweden tomorrow.

apetrynet commented:
Cool! Thanks!

On Thu, Apr 30, 2020, 11:51 Erik Johansson <notifications@github.com> wrote:

> @Eiken <https://github.com/Eiken>, I noticed via your PR that you use
> Hiero. Would you be able to give this plugin a spin?
>
> Sure thing, I'll get to it sometime early next week. Bank holiday in
> Sweden tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/284#issuecomment-621731942>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIW7YIAWUH7B5ARNSR7QADRPFC33ANCNFSM4FJBYAIA>
> .
>

apetrynet commented:
Just checked, and the Nuke Studio and Hiero share the same api and plugin structure. So the plugin works in both apps.
I might need to rename a few things to avoid confusion.

Eiken commented:
@apetrynet Do I need to compile opentime for this to work?

Tried just adding py-opentimelineio to PYTHONPATH and the plugin to HIERO_PLUGIN_PATH

Sidenote: What you say about nuke studio its correct. Same API with some stuff disabled from the normal nuke module when in Hiero.

Eiken commented:
traceback

```
Traceback (most recent call last):
  File "/usr/local/Nuke11.3v3/pythonextensions/site-packages/hiero/core/find_plugins.py", line 53, in loadPluginsFromFolder
    module = __import__( packageName )
  File "/path/OpenTimelineIO/contrib/opentimelineio_contrib/application_plugins/hiero/Python/StartupUI/otioimporter/__init__.py", line 31, in <module>
    from otioimporter.OTIOImport import load_otio
  File "/path/OpenTimelineIO/contrib/opentimelineio_contrib/application_plugins/hiero/Python/StartupUI/otioimporter/OTIOImport.py", line 38, in <module>
    import opentimelineio as otio
  File "/path/OpenTimelineIO/src/py-opentimelineio/opentimelineio/__init__.py", line 35, in <module>
    from . import (
  File "/path/OpenTimelineIO/src/py-opentimelineio/opentimelineio/opentime.py", line 1, in <module>
    from . _opentime import ( # noqa
ImportError: No module named _opentime
```

apetrynet commented:
Hi @Eiken ! 
Thanks for testing! You'll need to have a functioning build of OTIO to use it yes. So I think you would need to build OTIO and point PYTHONPATH to that location.

(I'm on linux so my knowledge is based on that only.)
I mostly followed the guide [here](https://opentimelineio.readthedocs.io/en/latest/tutorials/quickstart.html) to build OTIO.

Below are the steps I used to build the package.
When I do `ldd OpenTimelineIO/dist/lib/python2.7/site-packages/opentimelineio/_otio.so` I have no missing libraries. If you do, you might need to add `LD_LIBRARY_PATH=/path/otio/lib/` to your environment before running NS

``` bash
cd OpenTimelineIO
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX=../dist
make
cd ..
python setup.py install --cxx-install-root dist/ --prefix dist/
```
Thanks again for checking this out! Please let me know how it plays out.

Eiken commented:
pip installed latest version to our network share and added that to PYTHONPATH and I appear to be golden. Will get our editorial department to check it out.

apetrynet commented:
Haha! Yeah, that's way simpler.. Thanks! Looking forward to your feedback.

apetrynet commented:
Hi, @Eiken!
Have your editorial guys had a chance to look at the plugins?

Eiken commented:
Seems to work as expected. Tested some exporting and importing.

The Import menu was a little tricky to find tho when being lazy and not reading the README.
It would probably be good to see if it's possible to inject it into all other places where you can import XML and other formats as that's where users will start looking first. Especially in the main File menu.

apetrynet commented:
Thanks!
I didn't want to bee too intrusive so kept the menu injection to a minimal, but point taken. I'll see what I can do.

apetrynet commented:
Hmm. I might of course have done something wrong, but there seems to be a change upstream that causes the flake issues to appear in other areas than this plugin.

apetrynet commented:
@ssteinbach, do you think we can remove the "wip" label?

ssteinbach commented:
Yeah!  I'll take a look through the code.

ssteinbach commented on contrib/opentimelineio_contrib/application_plugins/hiero/Python/StartupUI/otioimporter/OTIOImport.py:
```diff
@@ -0,0 +1,536 @@
+                otio_item.out_offset.value
+            )
+
+        except Exception as e:
```
Is there a more specific exception you could listen for?  like, are these typically coming from OTIO or does Hiero have a particular exception this might raise?

apetrynet replied:
Oooh.. looks like I was a bit lazy there. I'll find the error and catch it.
Do you know any developers at the foundry? Would be great to get some feedback from them as well.

jhodges10 commented:
@ssteinbach I'll see if I can get someone at The Foundry to bite and take a look!

jhodges10 commented:
I mentioned this in Slack, but I'll say it here too. The folks at The Foundry are taking a look @apetrynet!

apetrynet commented:
Thanks!

On Thu, Jun 11, 2020, 20:28 Jeff <notifications@github.com> wrote:

> I mentioned this in Slack, but I'll say it here too. The folks at The
> Foundry are taking a look @apetrynet <https://github.com/apetrynet>!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PixarAnimationStudios/OpenTimelineIO/pull/284#issuecomment-642855848>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIW7YK3IIDIJWE6BKJEV7LRWEO3VANCNFSM4FJBYAIA>
> .
>

apetrynet commented:
@ssteinbach @jminor, Should this PR perhaps get linked to the [Interopertion](https://github.com/PixarAnimationStudios/OpenTimelineIO/projects/5) project?

apetrynet commented:
As discussed in TSC meeting I'll close this PR and maintain the plugin in a separate [repo](https://github.com/apetrynet/hiero-otio)
Plugin will be released on www.nukepedia.com shortly
